src/fields/mod.rs

```diff
@@ -0,0 +1,353 @@
+};
+use subtle::{Choice, ConstantTimeEq};
+
+pub trait FieldElement:
+    Debug
+    + Clone
+    + Copy
+    + ConstantTimeEq
+    + From<u64>
+    + Add<Output = Self>
+    + for<'a> Add<&'a Self, Output = Self>
+    + AddAssign
+    + Mul<Output = Self>
+    + for<'a> Mul<&'a Self, Output = Self>
+    + Neg<Output = Self>
+    + for<'a> TryFrom<&'a [u8], Error = anyhow::Error>
+    + Codec
+{
+    const NUM_BITS: u32;
+    const ZERO: Self;
+    const ONE: Self;
+
+    /// Number of bytes needed to represent a field element.
+    fn num_bytes() -> usize {
+        (Self::NUM_BITS as usize).div_ceil(8)
+    }
+
+    /// Project an integer into the field.
+    fn from_u128(value: u128) -> Self;
+
+    /// Test whether this element is zero.
+    fn is_zero(&self) -> Choice {
+        self.ct_eq(&Self::ZERO)
+    }
+}
+
+/// Field identifier. According to the draft specification, the encoding is of variable length ([1])
+/// but in the Longfellow implementation ([2]), they're always 3 bytes long.
+///
+/// [1]: https://datatracker.ietf.org/doc/html/draft-google-cfrg-libzk-00#section-7.2
+/// [2]: https://github.com/google/longfellow-zk/blob/902a955fbb22323123aac5b69bdf3442e6ea6f80/lib/proto/circuit.h#L309
+#[derive(Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord)]
+#[repr(u8)]
+pub enum FieldId {
+    /// The absence of a field, presumably if some circuit or proof has no subfield. This isn't
+    /// described in the specification (FieldID values start at 1) but is present in the Longfellow
+    /// implementation ([1]).
+    ///
+    /// [1]: https://github.com/google/longfellow-zk/blob/87474f308020535e57a778a82394a14106f8be5b/lib/proto/circuit.h#L55
+    None = 0,
+    /// NIST P256.
+    P256 = 1,
+    /// [`FieldP128`]
+    FP128 = 6,
+    // TODO: other field IDs as we need them
+}
+
+impl TryFrom<u8> for FieldId {
+    type Error = anyhow::Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(Self::None),
+            1 => Ok(Self::P256),
+            6 => Ok(Self::FP128),
+            _ => Err(anyhow!("unknown field ID")),
+        }
+    }
+}
+
+impl Codec for FieldId {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, anyhow::Error> {
+        let value = bytes
+            .read_u24::<LittleEndian>()
+            .context("failed to read u24")?;
+        let as_u8: u8 = value.try_into().context("decoded value too big for u8")?;
+        Self::try_from(as_u8)
+    }
+
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), anyhow::Error> {
+        bytes
+            .write_u24::<LittleEndian>(*self as u32)
+            .context("failed to write u24")
+    }
+}
+
+impl FieldId {
+    /// Returns the number of bytes occupied by the encoding of a field element of this ID.
+    pub fn encoded_length(&self) -> usize {
+        match self {
+            FieldId::None => 0,
+            FieldId::P256 => 32,
+            FieldId::FP128 => 16,
```

Comment:
I wonder if we should have these dispatch to `FieldElement::num_bytes()` instead of duplicating the constants.